eip/EIPS/eip-4337.md

```diff
@@ -271,23 +280,39 @@ In order to distinguish between them, there is a call to the NUMBER opcode (`blo
-7. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` results in an out-of-gas revert.
-8. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` to addresses with `EXTCODESIZE=0`.
-9. Any `GAS` opcode is followed immediately by one of { `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` }.
-10. `EXTCODEHASH` of every address accessed (by any opcode) does not change between first and second simulations of the op.
-11. If `op.initcode.length != 0` , allow only one `CREATE2` opcode call (in the validateUserOp block), otherwise forbid `CREATE2`.
-
-Since the wallet is allowed to access its own entry point deposit in order to top it up when needed, the client must know the storage slot in order to whitelist it. The entry point therefore implements the following view function:
-
-```c++
-function getSenderStorage(address sender) external view returns (uint256[] memory senderStorageCells)
-```
+2. The first (validateUserOp) call is allowed to access storage slot of the wallet itself (see Storage access by Slots, below), 
+   but only on contracts that are not part of the current bundle (that is, not in a paymaster or any other wallet)
+   Note that the wallet is allowed (and expected) to deposit to its own balance in the EntryPoint. That is allowed by this rule.
+3. The second (validatePaymasterUserOp), the paymaster may access storage slots of itself AND of the wallet, above.
+4. The third (validateUserOpSignature) view-call, the aggregator is allowed to access storage slots of itself and of the wallet. 
+5. Any `CALL` or `CALLCODE` during validation has `value=0`, except for the transfer from the wallet to the entry point.
+6. No `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` results in an out-of-gas revert.
+7. No referenced address (`CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL`, `EXTCODESIZE`, `EXTCODEHASH` to addresses with `EXTCODESIZE=0`.
+8. Any `GAS` opcode is followed immediately by one of { `CALL`, `DELEGATECALL`, `CALLCODE`, `STATICCALL` }.
+9. `EXTCODEHASH` of every address accessed (by any opcode) does not change between first and second simulations of the op.
+10. If `op.initcode.length != 0` , allow only one `CREATE2` opcode call (in the validateUserOp block), otherwise forbid `CREATE2`.
+
+#### Storage access by slots
+
+During validation, a contract `A` may reference storage of
+1. Slots of contract `A` address itself. 
+2. Slot `A` on any other address. 
+3. Slots of type `keccak256(A || X)` on any other address. (to cover `mapping(address => value)`, which is usually used for balance in ERC20 tokens). 
+4. Slots of type `keccak256(X || OWN)` on any other address, where OWN is some slot of the previous (third) type, or recursively of this (forth) type
+  (to cover `mapping(address ⇒ mapping(address ⇒ uint256)`) that are usually used for `allowances` in ERC20 tokens).
+NOTE: the total slots per wallet is limited to 30. above that, the UserOperation is rejected
```

Comment:
Why introduce a separate limit for storage access count? Why set it to `30`?
The slot is either valid or not, and there is gas limit.

Reply:
Yes.  See what I wrote about it in my previous review: https://github.com/eth-infinitism/account-abstraction/pull/121#discussion_r1017117460

If my back-of-an-envelope calculation is correct (please verify it), then gas limit should be enough.  An additional 50KB memory used during validation is unlikely to be an issue.  If it were in the order of MBs I'd think differently.  But in any case 30 is too low and prevents the use case I described above.

Reply:
removed the limit.